injectproxy/enforce_test.go

```diff
@@ -357,3 +366,533 @@ func TestEnforce(t *testing.T) {
+				},
+				{
+					`job=~"bar"`,
+					``,
+					true,
+				},
+				{
+					`job=~"fred"`,
+					``,
+					true,
+				},
+				{
+					`job=~"foo|fred"`,
+					`up{job="foo"}`,
+					false,
+				},
+				{
+					`job=~"foo|bar"`,
+					`up{job="foo"}`,
+					false,
+				},
+
+				// Not-regexp label selector in the expression.
+				{
+					`job!~""`,
+					`up{job="foo"}`,
+					false,
+				},
+				{
+					`job!~"foo"`,
+					``,
+					true,
+				},
+				{
+					`job!~"bar"`,
+					`up{job="foo"}`,
+					false,
+				},
+				{
+					`job!~"fred"`,
+					`up{job="foo"}`,
+					false,
+				},
+				{
+					`job!~"foo|fred"`,
+					``,
+					true,
+				},
+				{
+					`job!~"foo|bar"`,
+					``,
+					true,
+				},
+			},
+		},
+
+		// Not equal matcher enforcer.
+		{
+			enforcedMatcher: mustNewMatcher(labels.MatchNotEqual, "job", "foo"),
+			stcs: []subTestCase{
+				// No selector in the expression for the enforced label.
+				{
+					``,
+					`up{job!="foo"}`,
+					false,
+				},
+
+				// Equal label selector in the expression.
+				{
+					`job=""`,
+					`up{job!="foo",job=""}`,
+					false,
+				},
+				{
+					`job="foo"`,
+					``,
+					true,
+				},
+				{
+					`job="bar"`,
+					`up{job!="foo",job="bar"}`,
+					false,
+				},
+				{
+					`job="fred"`,
+					`up{job!="foo",job="fred"}`,
+					false,
+				},
+
+				// Not equal label selector in the expression.
+				{
+					`job!=""`,
+					`up{job!="",job!="foo"}`,
+					false,
+				},
+				{
+					`job!="foo"`,
+					`up{job!="foo"}`,
+					false,
+				},
+				{
+					`job!="bar"`,
+					`up{job!="bar",job!="foo"}`,
+					false,
+				},
+				{
+					`job!="fred"`,
+					`up{job!="foo",job!="fred"}`,
+					false,
+				},
+
+				// Regexp label selector in the expression.
+				{
+					`job=~""`,
+					`up{job!="foo",job=~""}`,
+					false,
+				},
+				{
+					// up{job!="foo",job=~"foo"} would return no result.
+					`job=~"foo"`,
+					``,
+					true,
+				},
+				{
+					`job=~"bar"`,
+					`up{job!="foo",job=~"bar"}`,
+					false,
+				},
+				{
+					`job=~"fred"`,
+					`up{job!="foo",job=~"fred"}`,
+					false,
+				},
+				{
+					`job=~"foo|fred"`,
+					`up{job!="foo",job=~"foo|fred"}`,
+					false,
+				},
+				{
+					`job=~"foo|bar"`,
+					`up{job!="foo",job=~"foo|bar"}`,
+					false,
+				},
+
+				// Not-regexp label selector in the expression.
+				{
+					`job!~""`,
+					`up{job!="foo",job!~""}`,
+					false,
+				},
+				{
+					`job!~"foo"`,
+					`up{job!="foo",job!~"foo"}`,
+					false,
+				},
+				{
+					`job!~"bar"`,
+					`up{job!="foo",job!~"bar"}`,
+					false,
+				},
+				{
+					`job!~"fred"`,
+					`up{job!="foo",job!~"fred"}`,
+					false,
+				},
+				{
+					`job!~"foo|fred"`,
+					`up{job!="foo",job!~"foo|fred"}`,
+					false,
+				},
+				{
+					`job!~"foo|bar"`,
+					`up{job!="foo",job!~"foo|bar"}`,
+					false,
+				},
+			},
+		},
+
+		// Regexp matcher enforcer.
+		{
+			enforcedMatcher: mustNewMatcher(labels.MatchRegexp, "job", "foo|bar"),
```

Comment:
there's some weird edge cases here that i'm not confident in because it depends so much on how the proxy's plumbing handles duplicate values. It's somewhat degenerate but I think it's possible but for an enforced matcher like `foo|foo` to be created. This would be disjoint with selectors like `job="bar"` and we *should* return errors in those cases since we are replacing things, right?